src/parser/retail/modules/items/dragonflight/enchants/StatProcEnchantAnalyzer.tsx

```diff
@@ -0,0 +1,102 @@
+import Spell from 'common/SPELLS/Spell';
+import { formatDuration, formatPercentage } from 'common/format';
+import QualityIcon from 'interface/QualityIcon';
+import { Options } from 'parser/core/EventSubscriber';
+import StatTracker from 'parser/shared/modules/StatTracker';
+import STAT, { getIcon, getName } from 'parser/shared/modules/features/STAT';
+import WeaponEnchantAnalyzer, { EnchantRank } from './WeaponEnchantAnalyzer';
+
+/**
+ * Export to reuse between the various secondary stat procc enchants.
+ */
+export const SECONDARY_STAT_WRIT_VALUES = {
+  1: 1185.67,
+  2: 1209.09,
+  3: 1394.51,
+} as const;
+
+export interface StatProcEnchantRank extends EnchantRank {
+  amount: number;
+}
+
+/**
+ * Abstraction to reuse the same code for tracking stats from proccing enchants such as
+ * writs and Sophic Devotion.
+ */
+abstract class StatProccEnchantAnalyzer extends WeaponEnchantAnalyzer<StatProcEnchantRank> {
+  static dependencies = {
+    ...WeaponEnchantAnalyzer.dependencies,
+    statTracker: StatTracker,
+  };
+  statTracker!: StatTracker;
+
+  /** The stat that the enchant provides on procc */
+  protected stat: STAT;
+  protected buff: Spell;
+
+  constructor(
+    stat: STAT,
+    enchantSpell: Spell,
+    buff: Spell,
+    ranks: StatProcEnchantRank[],
+    options: Options,
+  ) {
+    super(enchantSpell, ranks, options);
+    this.statTracker = (options as Options & { statTracker: StatTracker }).statTracker;
```

Comment:
If someone can tell me what's wrong when the DI system does not set `this.statTracker` but I can access it in `options`. Feels like this happens most of the time, but sometimes it works. I've tried to look in the debugger, but I cannot really follow

Reply:
its a quirk of the JS class system. superclasses cannot asssign properties to fields that are not in the superclass (even if they're in the subclass), which means that the `super` call cannot initialize most of the dependency fields. In general, this means that you can't use the DI modules via `this` until after the constructor call, at which point the `CombatLogParser` class actually assigns the values.

there's a new(ish) helper for this if you create a class by extending `Analyzer.withDependencies(deps)`, which will make typed access to the dependencies available on `this.deps`

Reply:
https://github.com/WoWAnalyzer/WoWAnalyzer/blob/dragonflight/src/analysis/retail/druid/restoration/modules/spells/GroveGuardians.tsx#L33 can see `withDependencies` used here

Reply:
That's very nifty! It doesn't seem to workout if I want to extend the classes I've created. I fiddled around with a mixin that is not a static method on the `Analyzer` class, and thus can add dependencies even if it's a second level descendant of `Analyzer`.

https://github.com/oBusk/WoWAnalyzer/commit/995c0b5ca89f7799713fd02646d582ebfc88a258

Do you think this could be something? It seems to work even if the middle generation adds their own `deps`.

Reply:
I am fine with `Analyzer.withDependencies` being a shorthand for this. seems reasonable.

ex:

```ts
class Analyzer extends Module {
  // ...
  static withDependencies = withDependencies.bind(null, Analyzer);
  // ...
}
```

Reply:
I've adopted this, please have a look @emallson